sdk/serialization/azure-json/src/main/java/com/azure/json/JsonWriter.java

```diff
@@ -248,22 +324,76 @@ private <T> JsonWriter writeArrayInternal(Iterable<T> array, WriteValueCallback<
-     * @throws NullPointerException If {@code valueWriterFunc} is null.
+     * @throws NullPointerException If {@code valueWriterFunc} is null or any key in the map is null.
      * @throws IOException If the JSON map fails to be written, either the start or end object or a key or value write.
      */
     public final <T> JsonWriter writeMap(Map<String, T> map, WriteValueCallback<JsonWriter, T> valueWriterFunc)
         throws IOException {
+        return writeMap(map, valueWriterFunc, false);
+    }
+
+    /**
+     * Writes a JSON map.
+     * <p>
+     * This API will begin by writing the start object ({@code &#123;}) followed by key-value fields in the map using
+     * the {@code valueWriterFunc} and finishing by writing the end object ({@code &#125;}).
+     * <p>
+     * If a value in the map is null and {@code skipNullValues} is true, that key-value pair will be skipped. Otherwise,
+     * handling of the null value will be based on {@code valueWriterFunc}. If {@code valueWriterFunc} skips null
+     * values, such as {@link #writeJson(JsonSerializable)}, {@link JsonToken#NULL} will be written for that value.
+     * <p>
+     * If {@code map} is null {@link JsonToken#NULL} will be written.
+     * <p>
+     * This API is used instead of {@link #writeMapField(String, Map, WriteValueCallback, boolean)} when the value needs
+     * to be written to the root of the JSON value, as an element in an array, or after a call to
+     * {@link #writeFieldName(String)}.
+     *
+     * @param map The map being written.
+     * @param valueWriterFunc The function that writes value of each key-value pair in the map.
+     * @param skipNullValues Whether null values in the map are skipped.
+     * @param <T> The value element type.
+     * @return The updated JsonWriter object.
+     * @throws NullPointerException If {@code valueWriterFunc} is null or any key in the map is null.
+     * @throws IOException If the JSON map fails to be written, either the start or end object or a key or value write.
+     */
+    public <T> JsonWriter writeMap(Map<String, T> map, WriteValueCallback<JsonWriter, T> valueWriterFunc,
+        boolean skipNullValues) throws IOException {
         Objects.requireNonNull(valueWriterFunc, "'valueWriterFunc' cannot be null.");
 
         if (map == null) {
             return writeNull();
         }
 
-        writeStartObject();
+        return writeMapInternal(null, map, valueWriterFunc, skipNullValues);
+    }
+
+    private <T> JsonWriter writeMapInternal(String fieldName, Map<String, T> map,
+        WriteValueCallback<JsonWriter, T> valueWriterFunc, boolean skipNullValues) throws IOException {
+        if (fieldName == null) {
+            writeStartObject();
+        } else {
+            writeStartObject(fieldName);
+        }
 
         for (Map.Entry<String, T> entry : map.entrySet()) {
+            T value = entry.getValue();
+            if (skipNullValues && value == null) {
```

Comment:
Can we invert this condition and just `writeNull()` here if `skipNullValues` is `false` and value is null? We don't have to keep track of the write context if we do that.

Reply:
What about keeping this the same and explicitly writing JSON null when the value is null? Inverting it would mean we need to track whether the field name should be written in multiple spots.